src/main/java/org/jenkinsci/remoting/protocol/ProtocolStack.java

```diff
@@ -760,7 +760,12 @@ public ProtocolStack<?> stack() {
          * Requests removal of this {@link ProtocolLayer} from the {@link ProtocolStack}
          */
         public void remove() {
-            removed = true;
+            stackLock.writeLock().lock();
+            try {
+                removed = true;
```

Comment:
interestingly `removed` is not `volatile`, could that be another cause of the issue?  

If we need the lock would that imply there are multiple threads at play and that this does also need to become volatile?

Reply:
I do not see a need for a memory barrier here but it could not hurt.